### DISCUSSION TITLE
VSCode and terragrunt

### DISCUSSION THREAD

**Discussion by sarkis** [gruntwork-io/knowledge-base#801] [Help] [answered]
Is there a recommended way to get get formatting for terragrunt enabled on vscode? I am trying to use a custom formatter, but this seems to cause some odd issues where on file save, the contents of the file vs the editor are different. Here is the snippet from settings:

```
"[hcl]": {
    "editor.defaultFormatter": "jkillian.custom-local-formatters",
  },
  "customLocalFormatters.formatters": [
      {
        "command": "terragrunt hclfmt -",
        "languages": ["hcl"]
      }
    ],
```

 Error I get randomly when using this:
 
 ```
Failed to save 'terragrunt.hcl': The content of the file is newer. Please compare your version with the file contents or overwrite the content of the file with your changes.
```

---

<ins datetime="2024-01-29T20:25:06Z">
  <p><a href="https://support.gruntwork.io/hc/requests/110728">Tracked in ticket #110728</a></p>
</ins>

**Comment by denis256** [comment, marked as answer]
Hi,
yes, now `terragrunt hclfmt` formats all files even before visual code save file, in my projects I use workaround script `format.sh` which save stdin to a temporary file, run formatting on it and read it back:

```
#!/bin/bash
temp_file=$(mktemp)
# save input to a temp file
cat > "$temp_file"
# format single file
terragrunt hclfmt --terragrunt-hclfmt-file "$temp_file"
# print rendered file
cat "$temp_file"
# clean temporary file
rm -f "$temp_file"
```

Configuration in `settings.json`:
```
{
  "editor.formatOnSave": true,
  "[hcl]": {
    "editor.defaultFormatter": "jkillian.custom-local-formatters",
  },
  "customLocalFormatters.formatters": [
    {
      "command": "./format.sh",
      "languages": [
        "hcl"
      ]
    }
  ]
}
```

**Reply by sarkis** [reply]
Started using it today and have not hit any of the issues with saving and filesystem having different content. Thank you for the solution.

**Reply by david-heward-unmind** [reply]
For me this saves the file with formatting but does not update the open editor 🤷

**Reply by laurynas** [reply]
You can do it without additional script now:
```json
{
    "editor.formatOnSave": true,
    "[hcl]": {
        "editor.defaultFormatter": "jkillian.custom-local-formatters",
    },
    "customLocalFormatters.formatters": [
        {
            "command": "terragrunt hcl format --stdin",
            "languages": [
                "hcl"
            ]
        }
    ]
}
```

**Comment by slaughtr** [comment]
For an alternative approach, I've been formatting files on save without any issue for 2+ years using [this extension](https://marketplace.visualstudio.com/items?itemName=emeraldwalk.RunOnSave) in conjunction with the following config in settings.json:

```
    "emeraldwalk.runonsave": {
        "commands": [
            {
                "cmd": "terragrunt hclfmt --terragrunt-hclfmt-file ${file}",
                "isAsync": true,
                "match": "\\.hcl$"
            }
        ]
    },
```

**Reply by dyegoe** [reply]
For those who are using newer terragrunt, use the same extension and update the code to:

```
"emeraldwalk.runonsave": {
    "commands": [
      {
        "cmd": "terragrunt hcl format --file=${file}",
        "isAsync": true,
        "match": "\\.hcl$"
      }
    ]
  },
```
